No-batch-dim support for ConvNd

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70506

This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

cc @ezyang @gchanan

jbschlosser committed:
No-batch-dim support for ConvNd

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser commented on test/test_modules.py:
```diff
@@ -417,9 +417,13 @@ def _test_gradients_helper(self, device, dtype, module_info, check):
             params = tuple(m.parameters())
 
-            # === Perform gradient check on the input_args ===
+            # === Lazy modules need to see an input to initialize params before gradcheck is run. ===
             input_args, input_kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
+            if issubclass(module_info.module_cls, torch.nn.modules.lazy.LazyModuleMixin):
+                with torch.no_grad():
+                    m(*input_args, **input_kwargs)
```
Tweak needed here for lazy modules.

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

albanD commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -731,60 +764,87 @@ Tensor _convolution_mode(
 
-  return at::convolution(input, weight, bias, stride, padding, dilation,
-                         true, output_padding, groups);
+  TORCH_CHECK(input_.dim() == 2 || input_.dim() == 3,
```
Why arent the transpose using the helper function?

jbschlosser replied:
Whoops, leftover from splitting the PR stack. Nice catch!

albanD commented on torch/nn/modules/conv.py:
```diff
@@ -233,8 +233,8 @@ class Conv1d(_ConvNd):
     Shape:
-        - Input: :math:`(N, C_{in}, L_{in})`
-        - Output: :math:`(N, C_{out}, L_{out})` where
+        - Input: :math:`(N, C_{in}, L_{in})` or :math:`(C_{in}, L_{in})`
```
We don't have a syntax to say that the `N` dimension is optional?

jbschlosser replied:
We discussed using `[N]` but went forward with this throughout as it's more explicit without the need for a legend.

I could be convinced otherwise, but we'd need to make consistent changes across modules and add the syntax to [this guide](https://github.com/pytorch/pytorch/wiki/torch.nn-Module-Documentation-Style-Guide).

Imo this doesn't look too bad with a description inline:
``` Input: :math:`([N], C_{in}, L_{in})`, where :math:`[N]` is an optional batch dimension ```

albanD replied:
I would personally vote for the `[N]` notation as it is shorter if we can use it consistently.

albanD commented on torch/nn/modules/conv.py:
```diff
@@ -1203,6 +1209,12 @@ def __init__(
         if bias:
             self.bias = UninitializedParameter(**factory_kwargs)
 
+    def _get_in_channels(self, input: Tensor) -> int:
```
nit: all of these are the same except the number of expected dimensions and function name. Should this be in a helper function like the c++ case?

albanD commented on torch/testing/_internal/common_modules.py:
```diff
@@ -314,6 +314,26 @@ def module_inputs_torch_nn_AvgPool1d(module_info, device, dtype, requires_grad,
 
+def module_inputs_torch_nn_ConvNd(module_info, device, dtype, requires_grad, **kwargs):
+    N = kwargs['N']
+    lazy = kwargs['lazy'] if 'lazy' in kwargs else False
```
Here and below:
```python
lazy = kwargs.get('lazy', False)
```

albanD reviewed (commented):
Looks mostly good.

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

albanD reviewed (approved):
SGTM

lint needs to be fixed though

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser committed:
Update on "No-batch-dim support for ConvNd"


This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

jbschlosser commented:
Labeled PR as BC-breaking because it was technically possible to (wrongly) pass e.g. 4D input / weight through `F.conv1d` before.
